src/pw/fpga/fft_fpga.c

```diff
@@ -0,0 +1,353 @@
+
+// host variables
+static cl_platform_id platform = NULL;
+static cl_device_id device = NULL;
+static cl_context context = NULL;
+static cl_program program = NULL;
+
+static cl_command_queue queue1 = NULL, queue2 = NULL, queue3 = NULL;
+static cl_command_queue queue4 = NULL, queue5 = NULL, queue6 = NULL;
+
+static cl_kernel fft_kernel = NULL, fft_kernel_2 = NULL;
+static cl_kernel fetch_kernel = NULL, transpose_kernel = NULL, transpose_kernel_2 = NULL;
+
+// Device memory buffers
+cl_mem d_inData, d_outData;
+
+// Global Variables
+static cl_int status = 0;
+static int fft_size[3] = {0,0,0};
+bool fft_size_changed = true;
+
+// Function prototypes
+bool init();
+void cleanup();
+void init_program(char *data_path);
+void queue_setup();
+void queue_cleanup();
+void fftfpga_run_3d(bool inverse, cmplx *c_in);
+
+// --- CODE -------------------------------------------------------------------
+
+int pw_fpga_initialize_(){
+   status = init();
+   return status;
+}
+
+void pw_fpga_final_(){
+   cleanup();
+}
+
+/******************************************************************************
+ * \brief  check whether FFT3d can be computed on the FPGA or not. This depends 
+ *         on the availability of bitstreams whose sizes are for now listed here 
+ *         The var fft_size_changed indicates whether a new binary needs to be 
+ *         loaded before execution. 
+ * \param  N - integer pointer to the size of the FFT3d
+ * \retval true if fft3d size supported
+ *****************************************************************************/
+bool pw_fpga_check_bitstream_(int N[3]){
+    // check the supported sizes
+    if( (N[0] == 16 && N[1] == 16 && N[2] == 16) ||
+        (N[0] == 32 && N[1] == 32 && N[2] == 32) ||
+        (N[0] == 64 && N[1] == 64 && N[2] == 64)  ){
+
+        // if previously used binary needs to be executed again, otherwise set
+        // the new size and load new binary
+        if( fft_size[0] == N[0] && fft_size[1] == N[1] && fft_size[2] == N[2] ){
+            fft_size_changed = false;
+        }
+        else{
+            fft_size[0] = N[0];
+            fft_size[1] = N[1];
+            fft_size[2] = N[2];
+            fft_size_changed = true;
+        }
+        return 1;
+    }
+    else{
+        return 0;
+    }
+}
+
+/******************************************************************************
+ * \brief   compute an in-place single precision complex 3D-FFT on the FPGA
+ * \param   data_path_len - length of the path to the data directory
+ * \param   data_path - path to the data directory 
+ * \param   direction : direction - 1/forward, otherwise/backward FFT3d
+ * \param   N   : integer pointer to size of FFT3d  
+ * \param   din : complex input/output single precision data pointer 
+ *****************************************************************************/
+void pw_fpga_fft3d_sp_(int data_path_len, char *data_path, int direction, int N[3], cmplx *din) {
+
+  data_path[data_path_len] = '\0';
```

Comment:
This assumes that buffer provided is bigger than the string length. If at all, shouldn't this be `data_path[data_path_len-1] = '\0'`?

Wouldn't it be easier to assume that Fortran calls the function with a `TRIM`med string ending with a `C_NULL_CHAR` instead? See also https://gcc.gnu.org/onlinedocs/gfortran/Interoperable-Subroutines-and-Functions.html